warehouse/query-core/src/main/java/datawave/query/util/sortedset/BufferedFileBackedSortedSet.java

```diff
@@ -58,25 +62,96 @@ public BufferedFileBackedSortedSet(BufferedFileBackedSortedSet<E> other) {
     
-    public BufferedFileBackedSortedSet(SortedSetFileHandlerFactory handlerFactory) {
-        this(null, DEFAULT_BUFFER_PERSIST_THRESHOLD, DEFAULT_MAX_OPEN_FILES, handlerFactory);
+    public BufferedFileBackedSortedSet(List<SortedSetFileHandlerFactory> handlerFactories) {
+        this(null, DEFAULT_BUFFER_PERSIST_THRESHOLD, DEFAULT_MAX_OPEN_FILES, DEFAULT_NUM_RETRIES, handlerFactories);
     }
     
-    public BufferedFileBackedSortedSet(Comparator<? super E> comparator, SortedSetFileHandlerFactory handlerFactory) {
-        this(comparator, DEFAULT_BUFFER_PERSIST_THRESHOLD, DEFAULT_MAX_OPEN_FILES, handlerFactory);
+    public BufferedFileBackedSortedSet(Comparator<? super E> comparator, List<SortedSetFileHandlerFactory> handlerFactories) {
+        this(comparator, DEFAULT_BUFFER_PERSIST_THRESHOLD, DEFAULT_MAX_OPEN_FILES, DEFAULT_NUM_RETRIES, handlerFactories);
     }
     
-    public BufferedFileBackedSortedSet(Comparator<? super E> comparator, int bufferPersistThreshold, int maxOpenFiles,
-                    SortedSetFileHandlerFactory handlerFactory) {
+    public BufferedFileBackedSortedSet(Comparator<? super E> comparator, int bufferPersistThreshold, int maxOpenFiles, int numRetries,
+                    List<SortedSetFileHandlerFactory> handlerFactories) {
         this.comparator = comparator;
-        this.handlerFactory = handlerFactory;
+        this.handlerFactories = handlerFactories;
         this.bufferPersistThreshold = bufferPersistThreshold;
+        this.numRetries = numRetries;
         this.maxOpenFiles = maxOpenFiles;
     }
     
+    private SortedSetFileHandler createFileHandler(SortedSetFileHandlerFactory handlerFactory) throws IOException {
+        boolean isCacheDirValid = false;
+        if (handlerFactory instanceof HdfsBackedSortedSet.SortedSetHdfsFileHandlerFactory) {
+            IvaratorCacheDir ivaratorCacheDir = ((HdfsBackedSortedSet.SortedSetHdfsFileHandlerFactory) handlerFactory).getIvaratorCacheDir();
+            
+            FsStatus fsStatus = null;
+            try {
+                fsStatus = ivaratorCacheDir.getFs().getStatus();
+            } catch (IOException e) {
+                log.warn("Unable to determine status of the filesystem: " + ivaratorCacheDir.getFs());
+            }
+            
+            // determine whether this fs is a good candidate
+            if (fsStatus != null) {
+                long availableStorageMB = fsStatus.getRemaining() / 0x100000L;
+                double availableStoragePercent = (double) fsStatus.getRemaining() / fsStatus.getCapacity();
+                
+                // if we are using less than our storage limit, the cache dir is valid
+                isCacheDirValid = availableStorageMB >= ivaratorCacheDir.getConfig().getMinAvailableStorageMB()
+                                && availableStoragePercent >= ivaratorCacheDir.getConfig().getMinAvailableStoragePercent();
+            }
+        } else {
+            isCacheDirValid = true;
```

Comment:
need to check available space for non-hdfs directories?

Reply:
We only ever create SortedSetHdfsFileHandlerFactory objects, so this case will never be reached in a deployment of DataWave.  However, for testing, there are tests which do NOT create SortedSetHdfsFileHandlerFactory objects, so we need to be able to handle that when we reach this logic.

Reply:
maybe this should be delegated then instead of implemented here?

Reply:
I'll create a validate method.

Reply:
done.